### PR TITLE
Simple assignments

### PR DESCRIPTION
**Pull request by Chestshire**
- This PR is to implement basic assignments for variables. Aim to have runnable code for a simple model. This works for one of the models but fails for another due to an issue with the state hierarchy. I will talk about what went wrong with one of the models.
- Another major change in this PR is the central controller. I rearranged the code in the central controller for basic commands and also to allow it to execute transitions in the states. We were not allowed to execute transitions without events before this. (@AidanCampbell738 @Simonliuwaterloo please be aware of this change)

Results:  
Dash model:
```
abstract sig Object { eats: set Object}
one sig Chicken, Farmer extends Object {}
conc state SimpleAssignments {
    conc state R{
        near: set Object
        far: set Object
        trans addFarmerChicken2near {
            do {
                near' = near + Farmer + Chicken
                far' = far - Farmer + Chicken
            }
        }
    }
}
```
Python code (only the transitions) 
```
# transition addFarmerChicken2near guard function
def _trans_addFarmerChicken2near_guard(self) -> bool:
    logging.log(GCLOG,f"Guard of [SimpleAssignments_R_addFarmerChicken2near] triggered.")
    return True

# transition addFarmerChicken2near execution function
def _trans_addFarmerChicken2near(self):
    logging.log(ACLOG,f"Execution of [SimpleAssignments_R_addFarmerChicken2near] triggered.")
    SS.SimpleAssignments_R_near = SS.SimpleAssignments_R_near + Farmer + Chicken
    SS.SimpleAssignments_R_far = SS.SimpleAssignments_R_far - Farmer + Chicken
    new_state = self._parent_state.SimpleAssignments_R_State(self._parent_state, self._root_state)
    self._parent_state.change_state(new_state)
```

Failed case:  
- Only one root conc state, so the `_parent_state` of the only state is `None`. However, it seems we don't have checks in the template file for this case so the code uses the `_parent_state` variable like it's not `None`.  

Dash Model:
```
abstract sig Object { eats: set Object}
one sig Chicken, Farmer extends Object {}
conc state SimpleAssignments {
    trans addFarmer2near {
        do {
            near' = near + Farmer
            far' = far - Farmer
        }
    }
}
```
Output: (line 280 to line 301)
```
class SimpleAssignments_State(State):

    def __init__(self, parent_state: State, root_state: State) -> None:
        super(SimpleAssignments_State, self).__init__(None, self) # init superclass
        # No substate for this state, so default state is None

        # initialize state transition dictionary
        self.transitions[self._trans_addFarmer2near_guard] = self._trans_addFarmer2near

    # transitions
    # transition addFarmer2near guard function
    def _trans_addFarmer2near_guard(self) -> bool:
        logging.log(GCLOG,f"Guard of [SimpleAssignments_addFarmer2near] triggered.")
        return True

    # transition addFarmer2near execution function
    def _trans_addFarmer2near(self):
        logging.log(ACLOG,f"Execution of [SimpleAssignments_addFarmer2near] triggered.")
        near = near + Farmer
        far = far - Farmer
        new_state = self._parent_state.SimpleAssignments_State(self._parent_state, self._root_state)
        self._parent_state.change_state(new_state)
```

### CODE REVIEW

File: org.alloytools.alloy.dash/src/main/java/ca/uwaterloo/watform/rapidDash/DashExprToPython.java [hunk 1]
```diff
@@ -18,7 +21,7 @@ a class used to translate expressions to Python code
  */
 public class DashExprToPython<ExprType> {
     private ExprType specialExpr;
-    private StringBuilder sb;
+    private Deque<StringBuilder> sbs;
```
Comment (Chestshire) [inline org.alloytools.alloy.dash/src/main/java/ca/uwaterloo/watform/rapidDash/DashExprToPython.java:24]: Considering we may see multiple lines of expressions, we need to split them and print them on each line. Hence, the stringbuilders, which are used to generate the expression strings, are stored in lists.

File: org.alloytools.alloy.application/src/main/java/templates/TestTemplateFile.vm [hunk 2]
```diff
@@ -378,42 +382,49 @@ class CentralController:
...
+        print("states: print the model's current states")
+        print("execute: take a random transition (if possible)")
+        print("quit: terminate the model\n")
+
+    def printStates(self):
+        print("\n" + type(self.concState).__name__ + " Current States:\n")
+        self.concState.printStates()
+
+    def execute(self):
+        self.concState.handleEvent()
```
Comment (Chestshire) [inline org.alloytools.alloy.application/src/main/java/templates/TestTemplateFile.vm:427]: Now we can type "execute" to execute a random transition.

File: org.alloytools.alloy.dash/src/main/java/ca/uwaterloo/watform/rapidDash/DashExprToPython.java [hunk 3]
```diff
@@ -70,35 +85,41 @@ private String genExpr(Expr node, int size) {
         // TODO: currently, the second parameter is redundant
 
         // check type, there are two types of expr
-        if (node instanceof ExprUnary) {
+        if (node instanceof  ExprList) {
```
Comment (Chestshire) [inline org.alloytools.alloy.dash/src/main/java/ca/uwaterloo/watform/rapidDash/DashExprToPython.java:88]: New stuff, expression list. When we see this, it means this expression has many independent expressions (probably linked with operation "AND").

File: org.alloytools.alloy.dash/src/main/java/ca/uwaterloo/watform/rapidDash/DashExprToPython.java [hunk 4]
```diff
@@ -291,7 +312,7 @@ private String BinaryOp2PythonOp(ExprBinary node){
             	} else {
             		res = this.genExpr(node.left, 1) + " = ";
             	}
-                
+                addParanthesis = false;
```
Comment (Chestshire) [inline org.alloytools.alloy.dash/src/main/java/ca/uwaterloo/watform/rapidDash/DashExprToPython.java:315]: No need to have a pair of parentheses for assignments.

**Review by RuomeiYan — APPROVED** [review]
LGTM, great job!

**Review by Simonliuwaterloo — APPROVED** [review]
LGTM

FYI I tried parsing the complete farmer example with this, an error (edu.mit.csail.sdg.ast.ExprVar cannot be cast to edu.mit.csail.sdg.ast.ExprBadJoin) is thrown due to this casting `ExprBadJoin badSubnode = (ExprBadJoin) badNode.right;`

**Review by AidanCampbell738 — APPROVED** [review]

**Pull request merged**